iidm/iidm-reducer/src/main/java/com/powsybl/iidm/reducer/DefaultNetworkReducer.java

```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    @Override
+    protected void reduce(TwoWindingsTransformer transformer) {
+        Terminal terminal1 = transformer.getTerminal1();
+        Terminal terminal2 = transformer.getTerminal2();
+        VoltageLevel vl1 = terminal1.getVoltageLevel();
+        VoltageLevel vl2 = terminal2.getVoltageLevel();
+
+        if (getFilter().accept(vl1)) {
+            replaceTransformerByLoad(transformer, vl1, terminal1);
+        } else if (getFilter().accept(vl2)) {
+            replaceTransformerByLoad(transformer, vl2, terminal2);
+        } else {
+            transformer.remove();
+            observers.forEach(o -> o.transformerRemoved(transformer));
+        }
+    }
+
+    @Override
+    protected void reduce(ThreeWindingsTransformer transformer) {
+        VoltageLevel vl1 = transformer.getLeg1().getTerminal().getVoltageLevel();
+        VoltageLevel vl2 = transformer.getLeg2().getTerminal().getVoltageLevel();
+        VoltageLevel vl3 = transformer.getLeg3().getTerminal().getVoltageLevel();
+        if (getFilter().accept(vl1) || getFilter().accept(vl2) || getFilter().accept(vl3)) {
+            throw new UnsupportedOperationException("Reduction of three-windings transformers is not supported");
+        } else {
+            transformer.remove();
+            observers.forEach(o -> o.transformerRemoved(transformer));
+        }
+    }
+
+    @Override
+    protected void reduce(HvdcLine hvdcLine) {
+        VoltageLevel vl1 = hvdcLine.getConverterStation1().getTerminal().getVoltageLevel();
+        VoltageLevel vl2 = hvdcLine.getConverterStation2().getTerminal().getVoltageLevel();
+        if (getFilter().accept(vl1) || getFilter().accept(vl2)) {
+            throw new UnsupportedOperationException("Reduction of HVDC lines is not supported");
+        } else {
+            hvdcLine.remove();
+            observers.forEach(o -> o.hvdcLineRemoved(hvdcLine));
+        }
+    }
+
+    private void reduce(Line line, VoltageLevel vl, Terminal terminal) {
+        if (options.isWithDanglingLines()) {
+            replaceLineByDanglingLine(line, vl, terminal);
+        } else {
+            replaceLineByLoad(line, vl, terminal);
+        }
+    }
+
+    private void replaceLineByLoad(Line line, VoltageLevel vl, Terminal terminal) {
+        Load load = replaceBranchByLoad(line, vl, terminal);
+        observers.forEach(o -> o.lineReduced(line, load));
+    }
+
+    private void replaceLineByDanglingLine(Line line, VoltageLevel vl, Terminal terminal) {
+        Branch.Side side = line.getSide(terminal);
+
+        DanglingLineAdder dlAdder = vl.newDanglingLine()
+                .setId(line.getId())
+                .setName(line.getName())
+                .setR(line.getR() / 2)
+                .setX(line.getX() / 2)
+                .setB(side == Branch.Side.ONE ? line.getB1() : line.getB2())
+                .setG(side == Branch.Side.ONE ? line.getG1() : line.getG2())
+                .setP0(terminal.getP())
```

Comment:
If p and q is not set you will have an exception coming from dangling line adder. Add a check with an understandable message?

Reply:
Done